Zend/zend_fiber.c

```diff
@@ -0,0 +1,633 @@
+/* fiber extension for PHP (c) 2017 Haitao Lv <php@lvht.net> */
+
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_closures.h"
+#include "zend_exceptions.h"
+#include "zend_fiber.h"
+#include "zend_interfaces.h"
+#include "zend_vm.h"
+
+static zend_class_entry *zend_ce_fiber;
+static zend_object_handlers zend_fiber_handlers;
+
+ZEND_DECLARE_MODULE_GLOBALS(fiber)
+
+static zend_object *zend_fiber_create(zend_class_entry *ce);
+static void zend_fiber_close(zend_fiber *fiber);
+static void fiber_interrupt_function(zend_execute_data *execute_data);
+
+static void (*orig_interrupt_function)(zend_execute_data *execute_data);
+
+static zend_op_array fiber_terminate_func;
+static zend_try_catch_element fiber_terminate_try_catch_array = {0,1,0,0};
+static zend_op fiber_terminate_op[2];
+
+static void zend_fiber_cleanup_unfinished_execution(zend_execute_data *execute_data) /* {{{ */
+{
+	if (execute_data->opline != execute_data->func->op_array.opcodes) {
+		/* -1 required because we want the last run opcode, not the next to-be-run one. */
+		uint32_t op_num = execute_data->opline - execute_data->func->op_array.opcodes - 1;
+
+		zend_cleanup_unfinished_execution(execute_data, op_num, 0);
+	}
+}
+/* }}} */
+
+static void zend_fiber_close(zend_fiber *fiber) /* {{{ */
+{
+	zend_vm_stack original_stack;
+	size_t original_stack_page_size;
+	zend_execute_data *execute_data;
+
+	if (UNEXPECTED(fiber->status == ZEND_FIBER_STATUS_INIT)) {
+		return;
+	}
+
+	original_stack = EG(vm_stack);
+	original_stack->top = EG(vm_stack_top);
+	original_stack->end = EG(vm_stack_end);
+	original_stack_page_size = EG(vm_stack_page_size);
+
+	EG(vm_stack_top) = fiber->stack->top;
+	EG(vm_stack_end) = fiber->stack->end;
+	EG(vm_stack) = fiber->stack;
+	EG(vm_stack_page_size) = fiber->stack_size;
```

Comment:
The saving/restoring logic seems to be repeated quite a lot. Wouldn't it be better to place the above in separate functions, just in case other members of the executor's globals need to be saved/restored in future? (I did something similar [here](https://github.com/tpunt/phactor/blob/6202b5ecbc61981557b6ca8c8642934cf0dcc34b/src/ph_context.c#L73).)

Reply:
The fiber is in its early stage and does not offer any Zend API now. Maybe we could introduce in the feature release if this patch were merged.